(#170) Add ability to remove gem from appraisal

jebentier opened the pull request:
## Purpose of PR
This is adding the ability to use the `remove_gem` declaration within an `appraise` block. By doing so the gem will be dropped from the generated Gemfile.  This is useful in cases where an appraisal require the removal of old dependencies that are still used by the mainline source.

## Example Usage
**Gemfile**
```ruby
gem 'rails', '~> 4.2'

group :test do
  gem 'rspec', '~> 4.0'
  gem 'test_after_commit'
end
```

**Appraisals**
```ruby
appraise 'rails-5' do
  gem 'rails', '~> 5.2'

  group :test do
    remove_gem 'test_after_commit'
  end
end
```

To produce:
**gemfiles/rails_5.gemfile**
```ruby
gem 'rails', '~> 5.2'

group :test do
  gem 'rspec', '~> 4.0'
end
```

Fixes #170

nickcharlton reviewed (approved):
Thanks! This looks good.

I think it would be useful to document this addition, do you think you could add it to the README? I'm thinking that perhaps the best approach here would be to add a new section for additional options like this

composerinteralia commented on spec/acceptance/appraisals_file_bundler_dsl_compatibility_spec.rb:
```diff
@@ -58,6 +60,7 @@
         end
 
         group :breakfast do
+          gem 'rotten_egg', :remove
```
Another API you might consider for this is:

```suggestion
          remove_gem 'rotten_egg'
```

I think this makes it a bit easier to parse visually, whereas with the `:remove` option you have to read to the end of the line to see whether the method is removing or adding something.

This also has the benefit of removing a conditional in the `gem` method.

jebentier replied:
I do like how this reads a lot better. I originally thought about doing this, but was hesitant to implement it this way due to Appraisal looking like it's trying to not add to the Bundler DSL, which adding the `remove_gem` command would be doing.  Though, I think the readability is worth the addition, so I'll go ahead and get this change worked in, along with some documentation in the README about how to utilize this like @nickcharlton mentioned.

nickcharlton replied:
Oh yeah, much nicer. Thanks for suggesting it, @composerinteralia!

composerinteralia commented on lib/appraisal/bundler_dsl.rb:
```diff
@@ -25,7 +25,11 @@ def run(&block)
 
     def gem(name, *requirements)
-      @dependencies.add(name, substitute_git_source(requirements))
+      if requirements.include?(:remove)
```
If we go with `remove_gem`, we wouldn't have a conditional in this method anymore. Instead we would introduce a new method:

```rb
def remove_gem(name)
  @dependencies.remove(name)
end
```

This also has the advantage of leaving off the `*requirements` argument when removing a gem, since we would have ignored most of those anyway.

jebentier replied:
This is much cleaner. I've gone ahead and implemented this just how you suggested. It's much simpler and easier to use.

composerinteralia reviewed (commented):
Nice! This is definitely something I would have used before if it existed.

jebentier commented:
@composerinteralia I've gone ahead and made the suggested edits to the new contract and updated the PR to reflect those changes.

@nickcharlton I've gone ahead and updated the `README` to include documentation for this new declaration

nickcharlton commented on README.md:
```diff
@@ -115,6 +115,45 @@ When you prefix a command with `appraisal`, the command is run with the
+`Appraisals` file.
+
+### Example Usage
+**Gemfile**
```
```suggestion
### Example Usage

**Gemfile**
```

nickcharlton replied:
Just a gap, nothing major just convention.

nickcharlton commented on README.md:
````diff
@@ -115,6 +115,45 @@ When you prefix a command with `appraisal`, the command is run with the
+end
+```
+
+**Appriasals**
````
```suggestion
**Appraisals**
```

nickcharlton commented on README.md:
```diff
@@ -115,6 +115,47 @@ When you prefix a command with `appraisal`, the command is run with the
+`Appraisals` file.
+
+### Example Usage
+### Example Usage
```
Oops, I think I did my change suggestion wrong! I don't think we wanted the double `### Example Usage`

jebentier replied:
😆  Fixed this, I should have looked at the suggestion closer before applying it. Thank you for catching it.

jebentier commented:
@composerinteralia would you have some time to re-review this in the coming days? I’m looking forward to getting this feature out so that we can use it at my company. If there’s anything that I can do to help with the release management of the gem as well, just let me know and I’d be happy to help.

nickcharlton commented:
Hi @jebentier! Sorry, that's on me. I was meaning to go ahead and merge this but it turned into a fairly busy week.

I'm going to merge this now and I'll take a look at doing a new release some time before this Friday. Thanks for your contribution!

jebentier commented:
@nickcharlton no worries at all, it’s been rather busy here as well. Thank you for accepting this contribution.

jebentier commented:
@nickcharlton any update on getting this released?

ColinDKelley commented:
@nickcharlton Any update on the release that will include this PR? It seems especially important since `master` has been documenting this new feature for 2 months now.

ColinDKelley commented:
@nickcharlton Any update on the release that will include this PR?

nickcharlton commented:
Thanks for your patience everyone — I just did this now and so you should see version 2.4.0 available now.
